File: user-sites/user_1003/brainfuck.py
#!/usr/bin/env python3
import sys

def run_bf(code):
    tape = [0] * 30000
    ptr = 0
    ip = 0
    loop_stack = []
    
    while ip < len(code):
        c = code[ip]
        if c == '>': ptr += 1
        elif c == '<': ptr -= 1
        elif c == '+': tape[ptr] = (tape[ptr] + 1) & 255
        elif c == '-': tape[ptr] = (tape[ptr] - 1) & 255
        elif c == '.': print(chr(tape[ptr]), end='')
        elif c == ',': tape[ptr] = ord(sys.stdin.read(1))
        elif c == '[':
            if tape[ptr] == 0:
                depth = 1
                while depth > 0:
                    ip += 1
                    if code[ip] == '[': depth += 1
                    elif code[ip] == ']': depth -= 1
            else:
                loop_stack.append(ip)
        elif c == ']':
            if tape[ptr] != 0:
                ip = loop_stack[-1]
            else:
                loop_stack.pop()
        ip += 1

if __name__ == '__main__':
    if len(sys.argv) > 1:
        run_bf(sys.argv[1])
    else:
        print("Usage: python3 brainfuck.py <code>")

File: user-sites/user_1003/brainfuck.c
#include <stdio.h>
#include <stdlib.h>

#define TAPE_SIZE 30000

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <brainfuck_code>\n", argv[0]);
        return 1;
    }
    
    char tape[TAPE_SIZE] = {0};
    char *ptr = tape;
    char *code = argv[1];
    char loop_stack[1000];
    int loop_sp = 0;
    
    for (int i = 0; code[i]; i++) {
        switch (code[i]) {
            case '>': ptr++; break;
            case '<': ptr--; break;
            case '+': (*ptr)++; break;
            case '-': (*ptr)--; break;
            case '.': putchar(*ptr); break;
            case ',': *ptr = getchar(); break;
            case '[':
                if (*ptr == 0) {
                    int depth = 1;
                    while (depth > 0) {
                        i++;
                        if (code[i] == '[') depth++;
                        else if (code[i] == ']') depth--;
                    }
                } else {
                    loop_stack[loop_sp++] = i;
                }
                break;
            case ']':
                if (*ptr != 0) {
                    i = loop_stack[loop_sp - 1];
                } else {
                    loop_sp--;
                }
                break;
        }
    }
    return 0;
}
